packages/main/src/RPA/Hubspot.py

```diff
@@ -0,0 +1,1706 @@
+        self,
+        object_type: str,
+        object_id: str,
+        id_property: Optional[str] = None,
+        properties: Optional[Union[str, List[str]]] = None,
+        associations: Optional[Union[str, List[str]]] = None,
+    ) -> Union[SimplePublicObject, SimplePublicObjectWithAssociations]:
+        self._require_authentication()
+
+        return self.hs.crm.objects.basic_api.get_by_id(
+            self._validate_object_type(object_type),
+            object_id,
+            properties=properties,
+            associations=(
+                [self._validate_object_type(obj) for obj in associations]
+                if associations
+                else None
+            ),
+            id_property=id_property,
+        )
+
+    @retry(
+        retry=retry_if_exception(_is_rate_limit_error),
+        stop=stop_after_attempt(10),
+        wait=wait_exponential(multiplier=2, min=0.1),
+        before_sleep=_before_sleep_log(),
+    )
+    def _get_object_by_batch(
+        self,
+        object_type: str,
+        object_id: List[str],
+        id_property: Optional[str] = None,
+        properties: Optional[Union[str, List[str]]] = None,
+    ) -> List[SimplePublicObject]:
+        self._require_authentication()
+        batched_ids = self._batch_batch_requests(object_id)
+        collected_responses = []
+        for i, batch in enumerate(batched_ids):
+            self.logger.debug(f"Executing batch index {i} of batch requests:\n{batch} ")
+            batch_reader = BatchReadInputSimplePublicObjectId(
+                properties=properties,
+                id_property=id_property,
+                inputs=[SimplePublicObjectId(o) for o in batch],
+            )
+            response = self.hs.crm.objects.batch_api.read(
+                self._singularize_object(self._get_custom_object_id(object_type)),
+                batch_read_input_simple_public_object_id=batch_reader,
+            )
+            if getattr(response, "num_errors", None):
+                if response.num_errors >= len(object_id):
+                    raise HubSpotBatchResponseError(
+                        "Batch API failed all items with the following "
+                        + f"errors for batch index {i}:\n{response.errors}"
+                    )
+                elif response.num_errors > 0:
+                    self.logger.warning(
+                        f"Batch API returned some errors for batch index {i}:\n"
+                        + str(response.errors)
+                    )
+            self.logger.debug(
+                f"Full results received for batch index {i}:\n{response.results}"
+            )
+            collected_responses.extend(response.results)
+
+        return collected_responses
+
+    @keyword
+    @retry(
+        retry=retry_if_exception(_is_rate_limit_error),
+        stop=stop_after_attempt(10),
+        wait=wait_exponential(multiplier=2, min=0.1),
+        before_sleep=_before_sleep_log(),
+    )
```

Comment:
Not sure if this should be retried as well since you're already retrying the underlying calls into `_get_object_by_batch` and `_get_object`.

Reply:
No, that's redundant, thanks!